### ISSUE TITLE
How to embed font into PDF file with number_pages()?

### ISSUE THREAD

**Issue by kauplan** [boazsegev/combine_pdf#137]
Hi,

I created page-numbered PDF file with `number_pages()` method successfully,
but I can't embed Helvetica font into PDF file.

example.rb:
```
require 'combine_pdf'

p CombinePDF::Fonts.fonts_list()
  #=> [:"Times-Roman", :"Times-Bold", :"Times-Italic", :"Times-BoldItalic",
  #    :Helvetica, :"Helvetica-Bold", :"Helvetica-BoldOblique", :"Helvetica-Oblique",
  #    :Courier, :"Courier-Bold", :"Courier-Oblique", :"Courier-BoldOblique",
  #    :Symbol, :ZapfDingbats]

#
# usage: ruby example.rb <file.pdf>
#
# sample pdf file: 12factor.epub_en.pdf
# https://github.com/bingohuang/12factor-gitbook/blob/master/download/12factor.epub_en.pdf
#
infile = ARGV[0]  or abort("ERROR: pdf file required.")
outfile = infile.sub(/\.pdf$/, '_numbered.pdf')
pdf = CombinePDF.load(infile)
pdf.number_pages(font: :Helvetica)
pdf.save(outfile)   # generates <file_numbered.pdf>
```

Document on `CombinePDF.register_font()` describes that `number_pages()` embeds font automatically, but not.

lib/combine_pdf/api.rb:
```
  :
  # registered fonts will remain in the library and will only be embeded in
  # PDF objects when they are used by PDFWriter objects (for example, for numbering pages).
  :
  def register_font(font_name, font_metrics, font_pdf_object, font_cmap = nil)
    Fonts.register_font font_name, font_metrics, font_pdf_object, font_cmap
  end
  :
```

How should I do to embed numbering font into PDF?

Thanks.

**Comment by boazsegev** [comment]
Hi @kauplan ,

The `Helvetica` font is one of the standard PDF fonts that should be available for every PDF reader, so there's no need to embed the font.

However, if you're using non ASCI characters, you will need to register a font.

To embed a font, you first need a PDF that includes the font.

The font is extracted from an existing PDF (you should have prior knowledge of the font's location in the PDF, which you can collect by testing a few options).

Once the font was extracted from an existing PDF, it should be registered in the new PDF.

You can [find an example using `CombinePDF.register_font_from_pdf_object`  here](https://github.com/boazsegev/combine_pdf_demo/blob/e16075410b211f2724bc41f35d8d2697e1d29270/pdf_controller.rb#L56-L67).

It's easier than it sounds and it worked well enough for one of my projects (which needed unicode support for the PDF fonts).

If you're running Rails, I would consider placing the code in one of the application's initializers.

Good luck!
   Bo.

**Comment by kauplan** [comment]
Very thanks for quick response.

> The Helvetica font is one of the standard PDF fonts that should be available for every PDF reader, so there's no need to embed the font.

I didn't know it. Thanks.

> To embed a font, you first need a PDF that includes the font.

I tried the following code, but it didn't work expectedly.

```
require 'combine_pdf'

filename = ARGV[0]  or abort "ERROR: pdf filename required."
pdf = CombinePDF.load(filename)
fonts = pdf.fonts(true)
p fonts   #=> []
```

Above code prints empty array, which contains no fonts. (ruby 2.5, combine_pdf 1.0.9, macOS)

Example PDF files:

* https://lise-henry.github.io/books/trpl2.pdf
* https://www.postgresql.org/files/documentation/pdf/9.0/postgresql-9.0-A4.pdf
* https://ufile.io/k6oqu (Created on macOS with Pages.app)

If there is wrong in above code, please let me know.

**Comment by boazsegev** [comment]
Are you sure these PDF files have any embedded fonts?

If they are using fonts such as `Helvetica` or `Courier`, they might not embed any fonts in the PDF (these are standard fonts).

**Comment by kauplan** [comment]
> Are you sure these PDF files have any embedded fonts?

Sure!
Here is a screenshot which proves that fonts are embedded in PDF file.
(Acrobat Reader DC, macOS)
https://imgur.com/ySieZpy

By the way, I defined `PDF#fonts2()` which is based on `PDF#fonts()`, and it seems to be able to retrieve fonts from PDF file above.

```ruby
require 'combine_pdf'

module CombinePDF
  class PDF
    def fonts2(limit_to_type0=false)
      fonts_found = []
      pages.each do |pg|
        font = pg[:Resources][:Font] || pg[:Resources][:referenced_object][:Font]
        next unless font
        font.values.each do |f|
          f = f[:referenced_object] if f[:referenced_object]
          next unless limit_to_type0 || f[:Subtype] == :Type01
          next unless f[:Type] == :Font && !fonts_found.include?(f)
          fonts_found << f
        end
      end
      fonts_found
    end
  end
end

filename = ARGV[0]  or abort "ERROR: pdf filename required."
pdf = CombinePDF.load(filename)
fonts = pdf.fonts2(true)
p fonts
```

I hope this helps you to address this issue.

**Comment by bassambarham** [comment]
Are French letters considered non-ascii?

When I add the word: réservation, it shows as: rˆ'servation

I've tried to add new fonts as suggested but the problem is still there.

Thanks

**Comment by boazsegev** [comment]
Yes, Some French letters are considered non ASCI.

It's fairly difficult to solve, since adding fonts isn't as simple as it sounds. CombinePDF does it's best to help, but it doesn't parse font data.

The way I solved it for Hebrew fonts was to:

1. Author a custom alphabet PDF using OpenOffice (at the time, now I'd do it using pages).

2. Add the fonts in that PDF, testing them one at a time to find the one that contained the Hebrew alphabet.

3. Hard-code the PDF choice in the alphabet PDF, using that PDF as a font source during startup.

There's an example [here](https://github.com/boazsegev/combine_pdf_demo/blob/286eca035b334a57c7c91837aa6d5f24c8399778/pdf_controller.rb#L56-L67) (this is the code I came up with after the process I described)... though the example uses lazy initialization (rather than loading the font during startup).
